// vite-plugin-clean/dist/index.cjs
'use strict';

const fs = require('fs');
const path = require('path');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e["default"] : e; }

const fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);
const path__default = /*#__PURE__*/_interopDefaultLegacy(path);

let options = {};
const { resolve, join } = path__default;
const { existsSync, readdirSync, statSync, unlinkSync, rmdirSync } = fs__default;
function cleanFiles(dirPath) {
  if (existsSync(dirPath)) {
    let files = readdirSync(dirPath);
    files.forEach((file) => {
      let path2 = join(dirPath, file);
      if (statSync(path2).isDirectory()) {
        cleanFiles(path2);
        rmdirSync(path2);
      } else {
        unlinkSync(path2);
      }
    });
  }
}
const cleanPlugin = (_opt) => {
  options = Object.assign({
    targetFiles: ["dist"]
  }, _opt);
  return {
    name: "vite-plugin-clean",
    enforce: "pre",
    apply: "build",
    buildStart: (config) => {
      const { targetFiles } = options;
      const rootDirectory = process.cwd();
      if (Array.isArray(targetFiles)) {
        for (let i = 0, len = targetFiles.length; i < len; i++) {
          cleanFiles(resolve(rootDirectory, targetFiles[i]));
        }
      } else {
        cleanFiles(resolve(rootDirectory, targetFiles));
      }
    }
  };
};

module.exports = cleanPlugin;
